# Support Vector Machine

In [1]:
import numpy as np
import pandas as pd
import cv2
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.svm import SVC

In [2]:
CSV_PATH = '../csv/'

We choose a dataset with picture's size equals to (64,64)

In [3]:
pictures = pd.read_csv(CSV_PATH + 'balanced_300_64.csv', header=None).to_numpy()
emotions = pd.read_csv(CSV_PATH + 'labels_300.csv', header=None).to_numpy().flatten()

In [4]:
pictures[:4]

array([[ 10.,  21.,  30., ...,  53.,  56.,  67.],
       [ 92.,  93.,  93., ..., 107., 118., 126.],
       [ 74.,  67.,  66., ...,  92.,  81.,  78.],
       [ 28.,  20.,  23., ..., 180., 179., 185.]])

In [5]:
emotions[:4]

array(['neutral', 'neutral', 'neutral', 'neutral'], dtype=object)

### Splitting

In [6]:
X_train, X_test, y_train, y_test = train_test_split(
    pictures,
    emotions,
    test_size=0.3,
    train_size=0.7,
    random_state=0
)

In [8]:
svm = SVC()

grid_params = {
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'gamma': ['scale', 'auto']
}

svm_tuned = GridSearchCV(svm, grid_params, cv=5, iid=False)

svm_tuned.fit(X_train, y_train)

print('SVM tuned parameters:', svm_tuned.best_params_)

print('Accuracy on train set:',svm_tuned.score(X_train, y_train))

print('Accuracy on test set:',svm_tuned.score(X_test, y_test))

SVM tuned parameters: {'gamma': 'scale', 'kernel': 'rbf'}
Accuracy on train set: 0.7763385146804835
Accuracy on test set: 0.4869215291750503


### Results

We get poor results even if they are better than the knn ones:
*  Train accuracy = 0.78
*  Test accuracy = 0.49

The model is overfitted since there is a big difference between the accuracies on the two sets.